packages/plugins/ens/src/index.js

```diff
@@ -90,17 +90,32 @@ class ENS {
     this.embark.registerActionForEvent("pipeline:generateAll:before", this.addArtifactFile.bind(this));
   }
 
-  getEnsConfig() {
-    return {
-      env: this.env,
-      registration: this.config.namesystemConfig.register,
-      registryAbi: this.ensConfig.ENSRegistry.abiDefinition,
-      registryAddress: this.ensConfig.ENSRegistry.deployedAddress,
-      registrarAbi: this.ensConfig.FIFSRegistrar.abiDefinition,
-      registrarAddress: this.ensConfig.FIFSRegistrar.deployedAddress,
-      resolverAbi: this.ensConfig.Resolver.abiDefinition,
-      resolverAddress: this.ensConfig.Resolver.deployedAddress
-    };
+  getEnsConfig(done) {
+    async.map(
+      this.config.namesystemConfig.dappConnection || this.config.contractsConfig.dappConnection,
+      (conn, next) => {
+      if (conn === '$EMBARK') {
+        return this.events.request('proxy:endpoint:get', (e, val) => {
+          next(e, val);
+        });
+      }
+      next(null, conn);
+    }, (err, connections) => {
```

Comment:
In response to your comment above, waiting on `proxy:endpoint:get` will wait for the proxy to be booted (as well as the blockchain started), so there really shouldn't be a timing issue **as long as** the ens module is started **after** the proxy in `cmd_controller/engine` (which it is: https://github.com/embark-framework/embark/blob/e7ea7db58e6481d5827273e7d600966e90f540f9/packages/core/core/src/engine.ts#L241-L250).

Two suggestions for this method:
1. Make it async - we don't need to rely on `async.map` and instead just use a `for` loop. This will also `await this.events.request2("proxy:endpoint:get")` before resolving.
2. Make it an (async) property  - because this is called multiple times to get the config, it might be a good idea that this is calculated once and uses backing variables to store the value. Because it's a property, the name should drop `get`, so it would be something like `ensConfig` or even just `config`:
```
async get config() {
  return (async () => {
      if (this._config) {
        return this._config;
      }
      const dappConnections = this.config.namesystemConfig.dappConnection || this.config.contractsConfig.dappConnection;
      const connections = [];
      for (conn of dappConnections) {
        if (conn === '$EMBARK') {
          connections.push(await this.events.request2("proxy:endpoint:get"));
        } else {
          connections.push(conn);
        }
      }
      this._config = {
        env: this.env,
        // NOTE: change `this.config` to `this.embarkConfig` in the constructor
        registration: this.embarkConfig.namesystemConfig.register,
        registryAbi: this.ensConfig.ENSRegistry.abiDefinition,
        registryAddress: this.ensConfig.ENSRegistry.deployedAddress,
        registrarAbi: this.ensConfig.FIFSRegistrar.abiDefinition,
        registrarAddress: this.ensConfig.FIFSRegistrar.deployedAddress,
        resolverAbi: this.ensConfig.Resolver.abiDefinition,
        resolverAddress: this.ensConfig.Resolver.deployedAddress,
        dappConnection: connections
      };
      return this._config;
    })();
}
```

Reply:
> In response to your comment above, waiting on proxy:endpoint:get will wait for the proxy to be booted (as well as the blockchain started), so there really shouldn't be a timing issue as long as the ens module is started after the proxy in cmd_controller/engine

Not entirely sure what comment you're referring too... but the issue was that, similar to `embarkjs/blockchain` we would replace `$EMBARK` with the proxy endpoint in  `addArtifactFile()` which is done right before the pipeline kicks in.

What happens however, is that we run `EmbarkJS.Names.setProvider()` in the VM, which means the code gets executed. By the time **that** happens, `$EMBARK` hasn't been replaced yet (because it happens right before pipeline kicks in, but after we set the provider in the console).

So the original idea was to make sure `$EMBARK` is replaced as early as possible, which I thought could/should be done in `Config.loadFiles()`, at the cost of making that function async. Here however, the problem is that, when `Config.loadFiles()` is called the first time, the blockchain node isn't spinned up yet, which means the proxy isn't booted up yet either, which results in `proxy:endpoint:get` requests going into nowhere.

IOW: we needed to find a place to request the proxy endpoint that's **before** the provider is set in the VM but after the proxy is loaded up. To preserve decoupling, I decided to replace `$EMBARK` whenever the configuration gets requested.

re: Your 2 suggestions: those are great! Will update the PR, thanks!

Reply:
Rolled back to the callback version. For reasoning: https://github.com/embark-framework/embark/pull/2110#issuecomment-562125729